pqarrow/arrowutils/sort.go

```diff
@@ -2,93 +2,303 @@ package arrowutils
+	}
+
+	build := array.NewInt32Builder(mem)
+	defer build.Release()
 
 	if r.NumRows() == 0 {
-		return indicesBuilder.NewInt64Array(), nil
+		return build.NewInt32Array(), nil
 	}
 	if r.NumRows() == 1 {
-		indicesBuilder.Append(0)
-		return indicesBuilder.NewInt64Array(), nil
+		build.Append(0)
+		return build.NewInt32Array(), nil
 	}
 
-	indices := make([]int64, r.NumRows())
-	// populate indices
-	for i := range indices {
-		indices[i] = int64(i)
+	ms := &multiColSorter{
+		indices:     make([]int32, r.NumRows()),
+		directions:  direction,
+		nullsFirst:  nullsFirst,
+		comparisons: make([]comparator, len(columns)),
 	}
-
-	switch c := r.Column(cols[0]).(type) {
-	case *array.Int64:
-		sort.Sort(orderedSorter[int64]{array: c, indices: indices})
-	case *array.String:
-		sort.Sort(orderedSorter[string]{array: c, indices: indices})
-	default:
-		return nil, fmt.Errorf("unsupported column type for sorting %T", c)
+	for i := range ms.indices {
+		ms.indices[i] = int32(i)
 	}
 
-	indicesBuilder.Reserve(len(indices))
-	for _, i := range indices {
-		indicesBuilder.Append(i)
+	for i, colIdx := range columns {
+		switch e := r.Column(colIdx).(type) {
+		case *array.Int64:
+			ms.comparisons[i] = &orderedSorter[int64]{array: e}
+		case *array.Float64:
+			ms.comparisons[i] = &orderedSorter[float64]{array: e}
+		case *array.String:
+			ms.comparisons[i] = &orderedSorter[string]{array: e}
+		case *array.Dictionary:
+			switch elem := e.Dictionary().(type) {
+			case *array.String:
+				ms.comparisons[i] = &orderedSorter[string]{
+					array: &stringDictionary{
+						dict: e,
+						elem: elem,
+					},
+				}
+			default:
+				return nil, fmt.Errorf("unsupported dictionary column type for sorting %T", e)
+			}
+		default:
+			return nil, fmt.Errorf("unsupported column type for sorting %T", e)
+		}
 	}
-
-	return indicesBuilder.NewInt64Array(), nil
+	sort.Sort(ms)
+	build.AppendValues(ms.indices, nil)
+	return build.NewInt32Array(), nil
 }
 
-// ReorderRecord reorders the given record's rows by the given indices.
-// This is a wrapper around compute.Take which handles the type castings.
-func ReorderRecord(ctx context.Context, r arrow.Record, indices arrow.Array) (arrow.Record, error) {
-	res, err := compute.Take(
-		ctx,
-		*compute.DefaultTakeOptions(),
-		compute.NewDatum(r),
-		compute.NewDatum(indices),
-	)
+// ReorderRecord uses indices to create a new record that is sorted according to
+// the indices array.
+func ReorderRecord(ctx context.Context, mem memory.Allocator, r arrow.Record, indices *array.Int32) (arrow.Record, error) {
+	ctx = compute.WithAllocator(ctx, mem)
+	// compute.Take doesn't support dictionaries. Use take on r when r does not have
+	// dictionary column.
+	var hasDictionary bool
+	for i := 0; i < int(r.NumCols()); i++ {
+		_, hasDictionary = r.Column(i).(*array.Dictionary)
+		if hasDictionary {
+			break
+		}
+	}
+	if !hasDictionary {
+		res, err := compute.Take(
+			ctx,
+			compute.TakeOptions{BoundsCheck: true},
```

Comment:
Isn't that the default? 
https://github.com/apache/arrow/blob/go/v14.0.0/go/arrow/compute/expression.go#L448

Reply:
Yes, the arrow api is pretty much broken on this regard `compute.DefaultTakeOptions` is a pointer while `compute.Take` accepts a non pointer. Doing pointer derefence for arguments in this case is just not okay.